File: solutions/c/acronym/1/acronym.c
#include "acronym.h"

char *abbreviate(const char *phrase){
    if(!phrase) return NULL;
    int len = strlen(phrase);
    if(len == 0){
        return NULL;
    }
    char *result = calloc(1, len);
    int pos = 1;
    bool valid = false;
    result[0] = phrase[0];
    for(int i = 1; i < len; i++){
        if(valid && isalpha(phrase[i])){
            result[pos++] = toupper(phrase[i]);
            valid = false;
        } else {
            switch(phrase[i]){
                case ' ':
                    valid = true;
                    break;
                case '-':
                    valid = true; 
                    break;
                case '_':
                    valid = true; 
                    break;
                default:
                    valid = false;
                    break;
            }
        }
    }
    return result;
}

File: solutions/zig/acronym/1/acronym.zig
const std = @import("std");
const mem = std.mem;

pub fn abbreviate(allocator: mem.Allocator, words: []const u8) mem.Allocator.Error![]u8 {
    const buff = try allocator.alloc(u8, words.len);
    
    var pos: u32 = 0;
    var wordstart = true;
    for(words[0..(words.len-1)]) |c| {
        switch(c){
            ' ', '_', '-' => {
                wordstart = true;
                continue;
            },
            else => {},
        }
        if(wordstart){
            buff[pos] = std.ascii.toUpper(c);
            pos += 1;
            wordstart = false;
        }
    }

    return try allocator.realloc(buff, pos);
}
